File: libagnos/csharp/src/Packers.cs
//////////////////////////////////////////////////////////////////////////////
// Part of the Agnos RPC Framework
//    http://agnos.sourceforge.net
//
// Copyright 2011, International Business Machines Corp.
//                 Author: Tomer Filiba (tomerf@il.ibm.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Net;
using System.Net.Sockets;
using System.IO;
using Agnos.Transports;


namespace Agnos
{
	public static class Packers
	{
		public const int CUSTOM_PACKER_ID_BASE = 2000;
	
		public class UnknownPackerId : ProtocolError
		{
			public UnknownPackerId(String message) : base(message) 
			{
			}
		}
		
		/////////////////////////////////////////////////////////////////////

		public abstract class AbstractPacker
		{
			public abstract void pack (object obj, ITransport transport);
			public abstract object unpack (ITransport transport);
			public abstract int getId ();

			protected static void _write (ITransport transport, byte[] buf)
			{
				try {
					transport.Write (buf, 0, buf.Length);
				} catch (IOException ex) {
					throw new EndOfStreamException ("write error", ex);
				} catch (SocketException ex) {
					throw new EndOfStreamException ("write error", ex);
				}
			}

			protected static void _read (ITransport transport, byte[] buf)
			{
				int total_got = 0;
				int got;
				
				try {
					while (total_got < buf.Length) {
						got = transport.Read (buf, total_got, buf.Length - total_got);
						total_got += got;
						if (got <= 0 && total_got < buf.Length) {
							throw new EndOfStreamException (String.Format("premature end of transport detected: " +
								"got={0}, expected={1}", total_got, buf.Length));
						}
					}
				} catch (IOException ex) {
					throw new EndOfStreamException ("read error", ex);
				} catch (SocketException ex) {
					throw new EndOfStreamException ("read error", ex);
				}
			}
		}

		/////////////////////////////////////////////////////////////////////

		public sealed class _Int8 : AbstractPacker
		{
			private byte[] buffer = new byte[1];

			internal _Int8 ()
			{
			}
			
			public override int getId()
			{
				return 1;
			}

			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					pack ((byte)0, transport);
				} else {
					pack ((byte)obj, transport);
				}
			}

			public void pack (byte val, ITransport transport)
			{
				buffer[0] = val;
				_write (transport, buffer);
			}

			public override object unpack (ITransport transport)
			{
				_read (transport, buffer);
				return buffer[0];
			}
		}

		public static readonly _Int8 Int8 = new _Int8 ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Bool : AbstractPacker
		{
			internal _Bool ()
			{
			}

			public override int getId()
			{
				return 2;
			}

			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					pack (false, transport);
				} else {
					pack ((bool)obj, transport);
				}
			}

			public void pack (bool val, ITransport transport)
			{
				if (val) {
					Int8.pack ((byte)1, transport);
				} else {
					Int8.pack ((byte)1, transport);
				}
			}

			public override object unpack (ITransport transport)
			{
				return ((byte)Int8.unpack (transport)) != 0;
			}
		}

		public static readonly _Bool Bool = new _Bool ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Int16 : AbstractPacker
		{
			private byte[] buffer = new byte[2];

			internal _Int16 ()
			{
			}
			
			public override int getId()
			{
				return 3;
			}

			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					pack ((short)0, transport);
				} else {
					pack ((short)obj, transport);
				}
			}

			public void pack (short val, ITransport transport)
			{
				buffer[0] = (byte)((val >> 8) & 0xff);
				buffer[1] = (byte)((val) & 0xff);
				_write (transport, buffer);
			}

			public override object unpack (ITransport transport)
			{
				_read (transport, buffer);
				return (short)((buffer[0] & 0xff) << 8 | (buffer[1] & 0xff));
			}
		}

		public static readonly _Int16 Int16 = new _Int16 ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Int32 : AbstractPacker
		{
			private byte[] buffer = new byte[4];

			internal _Int32 ()
			{
			}

			public override int getId()
			{
				return 4;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					pack ((int)0, transport);
				} else {
					pack ((int)obj, transport);
				}
			}

			public void pack (int val, ITransport transport)
			{
				buffer[0] = (byte)((val >> 24) & 0xff);
				buffer[1] = (byte)((val >> 16) & 0xff);
				buffer[2] = (byte)((val >> 8) & 0xff);
				buffer[3] = (byte)((val) & 0xff);
				_write (transport, buffer);
			}

			public override object unpack (ITransport transport)
			{
				_read (transport, buffer);
				return ((int)(buffer[0] & 0xff) << 24) | 
						((int)(buffer[1] & 0xff) << 16) | 
						((int)(buffer[2] & 0xff) << 8) | 
						((int)(buffer[3] & 0xff));
			}
		}

		public static readonly _Int32 Int32 = new _Int32 ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Int64 : AbstractPacker
		{
			private byte[] buffer = new byte[8];

			internal _Int64 ()
			{
			}

			public override int getId()
			{
				return 5;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					pack ((long)0, transport);
				} else {
					pack ((long)obj, transport);
				}
			}

			public void pack (long val, ITransport transport)
			{
				buffer[0] = (byte)((val >> 56) & 0xff);
				buffer[1] = (byte)((val >> 48) & 0xff);
				buffer[2] = (byte)((val >> 40) & 0xff);
				buffer[3] = (byte)((val >> 32) & 0xff);
				buffer[4] = (byte)((val >> 24) & 0xff);
				buffer[5] = (byte)((val >> 16) & 0xff);
				buffer[6] = (byte)((val >> 8) & 0xff);
				buffer[7] = (byte)(val & 0xff);
				_write (transport, buffer);
			}

			public override object unpack (ITransport transport)
			{
				_read (transport, buffer);
				return (((long)(buffer[0] & (uint)0xff)) << 56) | 
						(((long)(buffer[1] & (uint)0xff)) << 48) | 
						(((long)(buffer[2] & (uint)0xff)) << 40) | 
						(((long)(buffer[3] & (uint)0xff)) << 32) | 
						(((long)(buffer[4] & (uint)0xff)) << 24) | 
						(((long)(buffer[5] & (uint)0xff)) << 16) | 
						(((long)(buffer[6] & (uint)0xff)) << 8) | 
						(((long)(buffer[7] & (uint)0xff)));
			}
		}

		public static readonly _Int64 Int64 = new _Int64 ();

		/////////////////////////////////////////////////////////////////////

		public interface ISerializer
		{
			long store (object obj);
			object load (long id);
		}

        public class ObjRef : AbstractPacker
		{
			protected readonly ISerializer serializer;
            protected readonly int id;

			public ObjRef (int id, ISerializer serializer)
			{
				this.serializer = serializer;
				this.id = id;
			}

			public override int getId()
			{
				return id;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				Int64.pack (serializer.store (obj), transport);
			}

			public override object unpack (ITransport transport)
			{
				return serializer.load ((long)Int64.unpack (transport));
			}
		}

		/////////////////////////////////////////////////////////////////////

        public sealed class _Float : AbstractPacker
		{
			internal _Float ()
			{
			}

			public override int getId()
			{
				return 6;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					pack (0.0, transport);
				} else {
					pack ((double)obj, transport);
				}
			}

			public void pack (double val, ITransport transport)
			{
				Int64.pack (BitConverter.DoubleToInt64Bits (val), transport);
			}

			public override object unpack (ITransport transport)
			{
				return BitConverter.Int64BitsToDouble ((long)Int64.unpack (transport));
			}
		}

		public static readonly _Float Float = new _Float ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Buffer : AbstractPacker
		{
			internal _Buffer ()
			{
			}

			public override int getId()
			{
				return 7;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					Int32.pack (0, transport);
				} else {
					byte[] val = (byte[])obj;
					Int32.pack (val.Length, transport);
					_write (transport, val);
				}
			}

			public override object unpack (ITransport transport)
			{
				int length = (int)Int32.unpack (transport);
				byte[] buf = new byte[length];
				_read (transport, buf);
				return buf;
			}
		}

		public static readonly _Buffer Buffer = new _Buffer ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Date : AbstractPacker
		{
			internal _Date ()
			{
			}

			public override int getId()
			{
				return 8;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				long timestamp = ((DateTime)obj).ToUniversalTime().Ticks / 10;
				Int64.pack (timestamp, transport);
			}

			public override object unpack (ITransport transport)
			{
				long timestamp = (long)Int64.unpack(transport);
				return new DateTime(timestamp * 10, DateTimeKind.Utc);
			}
		}

		public static readonly _Date Date = new _Date ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Str : AbstractPacker
		{
            private readonly static UTF8Encoding utf8 = new UTF8Encoding();

			internal _Str ()
			{
			}

			public override int getId()
			{
				return 9;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					Buffer.pack (null, transport);
				} else {
					Buffer.pack (utf8.GetBytes ((string)obj), transport);
				}
			}

			public override object unpack (ITransport transport)
			{
				byte[] buf = (byte[])Buffer.unpack (transport);
				return utf8.GetString (buf);
			}
		}

		public static readonly _Str Str = new _Str ();

		/////////////////////////////////////////////////////////////////////

        public sealed class _Null : AbstractPacker
		{
			internal _Null ()
			{
			}

			public override int getId()
			{
				return 10;
			}
			
			public override void pack (object obj, ITransport transport)
			{
			}

			public override object unpack (ITransport transport)
			{
				return null;
			}
		}

		public static readonly _Null Null = new _Null ();

		/////////////////////////////////////////////////////////////////////

		public class ListOf<T> : AbstractPacker
		{
            protected readonly AbstractPacker type;
            protected readonly int id;

			public ListOf (int id, AbstractPacker type)
			{
				this.type = type;
				this.id = id;
			}

			public override int getId()
			{
				return id;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					Int32.pack (0, transport);
				} 
				else if (obj is T[]) {
					T[] val = (T[])obj;
					Int32.pack (val.Length, transport);
					foreach (T item in val) {
						type.pack (item, transport);
					}
				}
				else {
					IList<T> coll = (IList<T>)obj;
					Int32.pack (coll.Count, transport);
					foreach (T item in coll) {
						type.pack (item, transport);
					}
				}
			}

			public override object unpack (ITransport transport)
			{
				int length = (int)Int32.unpack (transport);
				List<T> lst = new List<T> (length);
				for (int i = 0; i < length; i++) {
					lst.Add ((T)(type.unpack (transport)));
				}
				return lst;
			}
		}

		public static readonly ListOf<byte> listOfInt8 = new ListOf<byte>(800, Int8);
		public static readonly ListOf<bool> listOfBool = new ListOf<bool>(801, Bool);
		public static readonly ListOf<short> listOfInt16 = new ListOf<short>(802, Int16);
		public static readonly ListOf<int> listOfInt32 = new ListOf<int>(803, Int32);
		public static readonly ListOf<long> listOfInt64 = new ListOf<long>(804, Int64);
		public static readonly ListOf<double> listOfFloat = new ListOf<double>(805, Float);
		public static readonly ListOf<byte[]> listOfBuffer = new ListOf<byte[]>(806, Buffer);
		public static readonly ListOf<DateTime> listOfDate = new ListOf<DateTime>(807, Date);
		public static readonly ListOf<string> listOfStr = new ListOf<string>(808, Str);
		
		/////////////////////////////////////////////////////////////////////

		public class SetOf<T> : AbstractPacker
		{
            protected readonly AbstractPacker type;
            protected readonly int id;

			public SetOf (int id, AbstractPacker type)
			{
				this.type = type;
				this.id = id;
			}

			public override int getId()
			{
				return id;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					Int32.pack (0, transport);
				} 
				else {
					ICollection<T> coll = (ICollection<T>)obj;
					Int32.pack (coll.Count, transport);
					foreach (T item in coll) {
						type.pack (item, transport);
					}
				}
			}

			public override object unpack (ITransport transport)
			{
				int length = (int)Int32.unpack (transport);
				HashSet<T> obj = new HashSet<T> ();
				for (int i = 0; i < length; i++) {
					obj.Add ((T)(type.unpack (transport)));
				}
				return obj;
			}
		}

		public static readonly SetOf<byte> setOfInt8 = new SetOf<byte>(820, Int8);
		public static readonly SetOf<bool> setOfBool = new SetOf<bool>(821, Bool);
		public static readonly SetOf<short> setOfInt16 = new SetOf<short>(822, Int16);
		public static readonly SetOf<int> setOfInt32 = new SetOf<int>(823, Int32);
		public static readonly SetOf<long> setOfInt64 = new SetOf<long>(824, Int64);
		public static readonly SetOf<double> setOfFloat = new SetOf<double>(825, Float);
		public static readonly SetOf<byte[]> setOfBuffer = new SetOf<byte[]>(826, Buffer);
		public static readonly SetOf<DateTime> setOfDate = new SetOf<DateTime>(827, Date);
		public static readonly SetOf<string> setOfStr = new SetOf<string>(828, Str);
		
		/////////////////////////////////////////////////////////////////////

		public class MapOf<K, V> : AbstractPacker
		{
            protected readonly AbstractPacker keytype;
            protected readonly AbstractPacker valtype;
            protected readonly int id;

			public MapOf (int id, AbstractPacker keytype, AbstractPacker valtype)
			{
				this.keytype = keytype;
				this.valtype = valtype;
				this.id = id;
			}

			public override int getId()
			{
				return id;
			}
			
			public override void pack (object obj, ITransport transport)
			{
				if (obj == null) {
					Int32.pack (0, transport);
				} else {
					IDictionary coll = (IDictionary)obj;
					Int32.pack (coll.Count, transport);
					foreach (DictionaryEntry item in coll) {
						keytype.pack (item.Key, transport);
						valtype.pack (item.Value, transport);
					}
				}
			}

			public override object unpack (ITransport transport)
			{
				int length = (int)Int32.unpack (transport);
				Dictionary<K, V> map = new Dictionary<K, V>(length);
				for (int i = 0; i < length; i++) {
					K k = (K)(keytype.unpack (transport));
					V v = (V)(valtype.unpack (transport));
					map.Add (k, v);
				}
				return map;
			}
		}

		public static readonly MapOf<int, int> mapOfInt32Int32 = new MapOf<int, int> (850, Int32, Int32);
		public static readonly MapOf<int, string> mapOfInt32Str = new MapOf<int, string> (851, Int32, Str);
		public static readonly MapOf<string, int> mapOfStrInt32 = new MapOf<string, int> (852, Str, Int32);
		public static readonly MapOf<string, string> mapOfStrStr = new MapOf<string, string> (853, Str, Str);

		// ////////////////////////////////////////////////////////////////////////

		public class HeteroMapPacker : AbstractPacker
		{
			protected readonly IDictionary<int, AbstractPacker> packersMap;
			protected int id;

			public HeteroMapPacker (int id, IDictionary<int, AbstractPacker> packersMap)
			{
				this.id = id;
				this.packersMap = packersMap;
			}

			public override int getId()
			{
				return id;
			}

			public override void pack (Object obj, ITransport transport)
			{
				HeteroMap map = (HeteroMap)obj;
				if (map == null) {
					Int32.pack (0, transport);
				} 
				else {
					Int32.pack (map.Count, transport);
					foreach (DictionaryEntry e in map) {
						AbstractPacker keypacker = map.getKeyPacker(e.Key);
						AbstractPacker valpacker = map.getValuePacker(e.Key);
						
						Int32.pack (keypacker.getId (), transport);
						keypacker.pack (e.Key, transport);
						
						Int32.pack (valpacker.getId (), transport);
						valpacker.pack (e.Value, transport);
					}
				}
			}

			public override Object unpack (ITransport transport)
			{
				int length = (int)Int32.unpack (transport);
				HeteroMap map = new HeteroMap (length);
				
				for (int i = 0; i < length; i++) {
					int keypid = (int)Int32.unpack (transport);
					AbstractPacker keypacker = getPacker (keypid);
					Object key = keypacker.unpack (transport);
					
					int valpid = (int)Int32.unpack (transport);
					AbstractPacker valpacker = getPacker (valpid);
					Object val = valpacker.unpack (transport);
					
					map.Add(key, keypacker, val, valpacker);
				}
				return map;
			}

			protected AbstractPacker getPacker (int id)
			{
				switch (id) {
				case 1:
					return Int8;
				case 2:
					return Bool;
				case 3:
					return Int16;
				case 4:
					return Int32;
				case 5:
					return Int64;
				case 6:
					return Float;
				case 7:
					return Buffer;
				case 8:
					return Date;
				case 9:
					return Str;
				case 10:
					return Null;
				case 800:
					return listOfInt8;
				case 801:
					return listOfBool;
				case 802:
					return listOfInt16;
				case 803:
					return listOfInt32;
				case 804:
					return listOfInt64;
				case 805:
					return listOfFloat;
				case 806:
					return listOfBuffer;
				case 807:
					return listOfDate;
				case 808:
					return listOfStr;				
				case 820:
					return setOfInt8;
				case 821:
					return setOfBool;
				case 822:
					return setOfInt16;
				case 823:
					return setOfInt32;
				case 824:
					return setOfInt64;
				case 825:
					return setOfFloat;
				case 826:
					return setOfBuffer;
				case 827:
					return setOfDate;
				case 828:
					return setOfStr;
				case 850:
					return mapOfInt32Int32;
				case 851:
					return mapOfInt32Str;
				case 852:
					return mapOfStrInt32;
				case 853:
					return mapOfStrStr;
				case 998:
					return builtinHeteroMapPacker;
				default:
					//if (id < CUSTOM_PACKER_ID_BASE) {
					//	throw new UnknownPackerId ("packer id too low" + id);
					//}
					if (packersMap == null) {
						throw new UnknownPackerId ("unknown packer id " + id);
					}
					AbstractPacker packer;
					if (!packersMap.TryGetValue(id, out packer)) {
						throw new UnknownPackerId ("unknown packer id " + id);
					}
					return packer;
				}
			}
		}

		public static readonly HeteroMapPacker builtinHeteroMapPacker = new HeteroMapPacker (998, null);
	}
	
	
}

File: libagnos/cpp/src/packers.cpp
//////////////////////////////////////////////////////////////////////////////
// Part of the Agnos RPC Framework
//    http://agnos.sourceforge.net
//
// Copyright 2011, International Business Machines Corp.
//                 Author: Tomer Filiba (tomerf@il.ibm.com)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//////////////////////////////////////////////////////////////////////////////

#include "packers.hpp"
#include <boost/detail/endian.hpp>

#if defined(BOOST_BIG_ENDIAN)
#	ifndef ntohs
#		define ntohs(n) (n)
#		define htons(n) (n)
#	endif
#	ifndef ntohl
#		define ntohl(n) (n)
#		define htonl(n) (n)
#	endif
#	ifndef ntohll
#		define ntohll(n) (n)
#		define htonll(n) (n)
#	endif
#elif defined(BOOST_LITTLE_ENDIAN)

static inline int16_t _bswap_16(int16_t n)
{
	return ((n & 0x00ffu) << 8) | ((n & 0xff00u) >> 8);
}

static inline int32_t _bswap_32(int32_t n)
{
	return ((n & 0x000000fful) << 24) | ((n & 0x0000ff00ul) << 8) | ((n
	        & 0x00ff0000ul) >> 8) | ((n & 0xff000000ul) >> 24);
}

static inline int64_t _bswap_64(int64_t n)
{
	return ((n & 0xff00000000000000ull) >> 56) | ((n & 0x00ff000000000000ull)
	        >> 40) | ((n & 0x0000ff0000000000ull) >> 24) | ((n
	        & 0x000000ff00000000ull) >> 8) | ((n & 0x00000000ff000000ull) << 8)
	        | ((n & 0x0000000000ff0000ull) << 24)
	        | ((n & 0x000000000000ff00ull) << 40)
	        | ((n & 0x00000000000000ffull) << 56);
}

#	ifndef ntohs
#		define ntohs(n)  _bswap_16(n)
#		define htons(n)  _bswap_16(n)
#	endif
#	ifndef ntohl
#		define ntohl(n)  _bswap_32(n)
#		define htonl(n)  _bswap_32(n)
#	endif
#	ifndef ntohll
#		define ntohll(n) _bswap_64(n)
#		define htonll(n) _bswap_64(n)
#	endif
#else
#	error "unknown machine endianity"
#endif // BOOST_BYTE_ORDER

#define IPACKER_SIMPLE_IMPL(CLS, NAME) \
	CLS::CLS() {} \
	int32_t CLS::get_id() const \
	{ \
		return _id; \
	} \
	void CLS::pack_any(const any& obj, ITransport& transport) const \
	{ \
		if (obj.type() == typeid(shared_ptr<data_type>)) { \
			shared_ptr<data_type> tmp = any_cast< shared_ptr<data_type> >(obj); \
			pack(*tmp, transport); \
		} \
		else { \
			pack(any_cast<data_type>(obj), transport); \
		} \
	} \
	any CLS::unpack_any(ITransport& transport) const \
	{ \
		data_type tmp; \
		unpack(tmp, transport); \
		return tmp; \
	} \
	any CLS::unpack_shared(ITransport& transport) const \
	{ \
		shared_ptr<data_type> obj(new data_type()); \
		unpack(*obj, transport); \
		return obj; \
	} \
	void CLS::pack(shared_ptr<CLS::data_type> obj, ITransport& transport) \
	{ \
		pack(*obj, transport); \
	} \
	void CLS::unpack(shared_ptr<CLS::data_type>& obj, ITransport& transport) \
	{ \
		obj.reset(new data_type()); \
		unpack(*obj, transport); \
	} \
	CLS NAME;

//////////////////////////////////////////////////////////////////////////////

namespace agnos
{
	namespace packers
	{
		inline static void _write(ITransport& transport, const char * buf,
		        size_t size)
		{
			transport.write(buf, size);
		}

		inline static void _read(ITransport& transport, char * buf, size_t size)
		{
			size_t total_got = 0;
			while (total_got < size) {
				size_t got = transport.read(buf + total_got, size - total_got);
				total_got += got;
				if (got <= 0 && total_got < size) {
					throw PackerError("unexpected EOF");
				}
			}
		}

		//////////////////////////////////////////////////////////////////////

		int32_t VoidPacker::get_id() const
		{
			throw PackerError("VoidPacker");
		}
		void VoidPacker::pack_any(const any& obj, ITransport& transport) const
		{
			throw PackerError("VoidPacker");
		}
		any VoidPacker::unpack_any(ITransport& transport) const
		{
			throw PackerError("VoidPacker");
		}
		any VoidPacker::unpack_shared(ITransport& transport) const
		{
			throw PackerError("VoidPacker");
		}

		VoidPacker void_packer;

		//////////////////////////////////////////////////////////////////////

		shared_ptr<_NullObj> NullObj;

		int32_t NullPacker::get_id() const
		{
			return 10;
		}
		void NullPacker::pack_any(const any& obj, ITransport& transport) const
		{
		}
		any NullPacker::unpack_any(ITransport& transport) const
		{
			return NullObj;
		}
		any NullPacker::unpack_shared(ITransport& transport) const
		{
			return NullObj;
		}

		NullPacker null_packer;

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(Int8Packer, int8_packer)

		void Int8Packer::pack(const int8_t& obj, ITransport& transport)
		{
			_write(transport, (const char *) &obj, sizeof(obj));
		}

		void Int8Packer::unpack(int8_t& obj, ITransport& transport)
		{
			_read(transport, (char *) &obj, sizeof(obj));
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(BoolPacker, bool_packer)

		void BoolPacker::pack(const bool& obj, ITransport& transport)
		{
			int8_t tmp = obj ? 1 : 0;
			Int8Packer::pack(tmp, transport);
		}

		void BoolPacker::unpack(bool& obj, ITransport& transport)
		{
			int8_t tmp;
			Int8Packer::unpack(tmp, transport);
			obj = (tmp != 0);
		}

		void BoolPacker::unpack(std::_Bit_reference obj, ITransport& transport)
		{
			bool tmp;
			unpack(tmp, transport);
			obj = tmp;
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(Int16Packer, int16_packer)

		void Int16Packer::pack(const int16_t& obj, ITransport& transport)
		{
			int16_t tmp = htons(obj);
			_write(transport, (const char *) &tmp, sizeof(tmp));
		}

		void Int16Packer::unpack(int16_t& obj, ITransport& transport)
		{
			int16_t tmp;
			_read(transport, (char *) &tmp, sizeof(tmp));
			obj = ntohs(tmp);
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(Int32Packer, int32_packer)

		void Int32Packer::pack(const int32_t& obj, ITransport& transport)
		{
			int32_t tmp = htonl(obj);
			_write(transport, (const char*) &tmp, sizeof(tmp));
		}
		void Int32Packer::unpack(int32_t& obj, ITransport& transport)
		{
			int32_t tmp;
			_read(transport, (char*) &tmp, sizeof(tmp));
			obj = ntohl(tmp);
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(Int64Packer, int64_packer)

		void Int64Packer::pack(const int64_t& obj, ITransport& transport)
		{
			int64_t tmp = htonll(obj);
			transport.write((const char*) &tmp, sizeof(tmp));
		}
		void Int64Packer::unpack(int64_t& obj, ITransport& transport)
		{
			int64_t tmp;
			transport.read((char*) &tmp, sizeof(tmp));
			obj = ntohll(tmp);
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(FloatPacker, float_packer)

		void FloatPacker::pack(const double& obj, ITransport& transport)
		{
			int64_t tmp = *(reinterpret_cast<const int64_t*>(&obj));
			Int64Packer::pack(tmp, transport);
		}

		void FloatPacker::unpack(double& obj, ITransport& transport)
		{
			int64_t tmp;
			Int64Packer::unpack(tmp, transport);
			obj = *(reinterpret_cast<double*>(&tmp));
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(BufferPacker, buffer_packer)

		void BufferPacker::pack(const string& obj, ITransport& transport)
		{
			Int32Packer::pack(obj.size(), transport);
			transport.write(obj.data(), obj.size());
		}
		void BufferPacker::unpack(string& obj, ITransport& transport)
		{
			int32_t size;
			Int32Packer::unpack(size, transport);
			obj.resize(size);
			transport.read(const_cast<char*> (obj.data()), size);
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(StringPacker, string_packer)

		void StringPacker::pack(const string& obj, ITransport& transport)
		{
			BufferPacker::pack(obj, transport);
		}
		void StringPacker::unpack(string& obj, ITransport& transport)
		{
			BufferPacker::unpack(obj, transport);
		}

		//////////////////////////////////////////////////////////////////////

		IPACKER_SIMPLE_IMPL(DatePacker, date_packer)

		const int64_t microsecs_from_epoch = 44148153600000000ll;
		static datetime mintime = boost::posix_time::from_iso_string(
		        std::string("14000101T000000"));

		void DatePacker::pack(const datetime& obj, ITransport& transport)
		{
			timespan dur = obj - mintime;
			int64_t microsecs = dur.total_microseconds() + microsecs_from_epoch;
			Int64Packer::pack(microsecs, transport);
		}

		void DatePacker::unpack(datetime& obj, ITransport& transport)
		{
			int64_t val;
			Int64Packer::unpack(val, transport);
			val -= microsecs_from_epoch;

			// urrgh, boost::posix_time::microseconds only accepts int32!
			// need to do it the hard way
			timespan dur = boost::posix_time::microseconds(val % 1000000);
			val /= 1000000;
			dur += boost::posix_time::seconds(val % 60);
			val /= 60;
			dur += boost::posix_time::minutes(val % 60);
			val /= 60;
			dur += boost::posix_time::hours(val);

			obj = mintime + dur;
		}

		//////////////////////////////////////////////////////////////////////

		ListPacker<int8_t, 800> list_of_int8_packer(int8_packer);
		ListPacker<bool, 801> list_of_bool_packer(bool_packer);
		ListPacker<int16_t, 802> list_of_int16_packer(int16_packer);
		ListPacker<int32_t, 803> list_of_int32_packer(int32_packer);
		ListPacker<int64_t, 804> list_of_int64_packer(int64_packer);
		ListPacker<double, 805> list_of_float_packer(float_packer);
		ListPacker<string, 806> list_of_buffer_packer(buffer_packer);
		ListPacker<datetime, 807> list_of_date_packer(date_packer);
		ListPacker<string, 808> list_of_string_packer(string_packer);

		SetPacker<int8_t, 820> set_of_int8_packer(int8_packer);
		SetPacker<bool, 821> set_of_bool_packer(bool_packer);
		SetPacker<int16_t, 822> set_of_int16_packer(int16_packer);
		SetPacker<int32_t, 823> set_of_int32_packer(int32_packer);
		SetPacker<int64_t, 824> set_of_int64_packer(int64_packer);
		SetPacker<double, 825> set_of_float_packer(float_packer);
		SetPacker<string, 826> set_of_buffer_packer(buffer_packer);
		SetPacker<datetime, 827> set_of_date_packer(date_packer);
		SetPacker<string, 828> set_of_string_packer(string_packer);

		MapPacker<int32_t, int32_t, 850> map_of_int32_int32_packer(int32_packer, int32_packer);
		MapPacker<int32_t, string, 851> map_of_int32_string_packer(string_packer, int32_packer);
		MapPacker<string, int32_t, 852> map_of_string_int32_packer(int32_packer, string_packer);
		MapPacker<string, string, 853> map_of_string_string_packer(string_packer, string_packer);

	}
}
